### PR TITLE
Add MaybeUninit array transpose impls

### LINKED ISSUE #110: Add array transpose From implementations: `[MaybeUninit<T>; N]` <-> `MaybeUninit<[T; N]>`
## Problem statement

The primary use case for `MaybeUninit` is to initialize arrays of stuff (typically u8s). This is currently poorly supported, requiring the user to enable unstable features (e.g. `uninit_array`) or immediately mark an array as `assume_init` which is very confusing to beginners.

## Motivation, use-cases

```rust
let data = [MaybeUninit::<u8>::uninit(); 2048]; // This could instead be MaybeUninit::<[u8; 2048]>::uninit(), but now you need to use raw pointers to write the data
for ... {
  data.write(...);
}
let data: [u8; 2048] = UH_OH;
```

UH_OH can be replaced with a `transmute` (Requires unsafe. Also is that actually safe? It is, but no one wants to spend hours digging around to figure that out.) or with `array_assume_init` (Requires nightly).

Having a `From` implementation between `[MaybeUninit<T>; N]` and `MaybeUninit<[T; N]>` makes it easy to move between filling an uninitialized array and then marking it as initialized.

## Solution sketches

See https://github.com/rust-lang/rust/pull/102023

## Links and related work

https://github.com/rust-lang/rust/pull/88837

Other proposals have suggested implementing `Index{,Mut}`. While I think that's probably a slightly nicer API in its limited scope, users are very quickly going to be sad when they see that their favorite `MaybeUninit` method isn't available on `[MaybeUninit<T>; N]`. Eventually that will lead to API duplication as `[MaybeUninit<T>; N]` methods are added to mirror the `MaybeUninit` ones. `[MaybeUninit<T>; N]` <-> `MaybeUninit<[T; N]>` conversion avoids this entirely.

### PR DESCRIPTION
**Pull request by SUPERCILEX**
See discussion in https://github.com/rust-lang/rust/pull/101179 and https://github.com/rust-lang/rust/issues/96097. I believe this solution offers the simplest implementation with minimal future API regret. Closes https://github.com/rust-lang/libs-team/issues/110.

@RalfJung mind doing a correctness review?

### CODE REVIEW

**Comment by SUPERCILEX** [comment]
rustbot label +T-libs-api -T-libs

**Comment by thomcc** [comment]
I'm definitely in favor of adding this, but I think it needs an ACP.

**Comment by SUPERCILEX** [comment]
@thomcc Sounds good, done: https://github.com/rust-lang/libs-team/issues/110

**Comment by SUPERCILEX** [comment]
Hmmm, no idea how those failures are related. Will ignore for now.

**Comment by RalfJung** [comment]
Yeah that's a strange error, can you rebase and force push?

File: src/test/ui/suggestions/issue-71394-no-from-impl.stderr [hunk 1]
```diff
@@ -7,6 +7,7 @@ LL |     let _: &[i8] = data.into();
    |                    the trait `From<&[u8]>` is not implemented for `&[i8]`
    |
    = help: the following other types implement trait `From<T>`:
+             <[MaybeUninit<T>; N] as From<MaybeUninit<[T; N]>>>
```
Comment (RalfJung) [inline src/test/ui/suggestions/issue-71394-no-from-impl.stderr:10]: This is unfortunate but I guess unavoidable with our current scheme...

File: library/core/src/mem/maybe_uninit.rs [hunk 2]
```diff
@@ -117,28 +117,22 @@ use crate::slice;
...
 ///
 /// let data = {
-///     // Create an uninitialized array of `MaybeUninit`. The `assume_init` is
-///     // safe because the type we are claiming to have initialized here is a
-///     // bunch of `MaybeUninit`s, which do not require initialization.
-///     let mut data: [MaybeUninit<Vec<u32>>; 1000] = unsafe {
-///         MaybeUninit::uninit().assume_init()
-///     };
+///     // Create an uninitialized array of `MaybeUninit`.
+///     let mut data = [MaybeUninit::uninit(); 1000];
```
Comment (RalfJung) [inline library/core/src/mem/maybe_uninit.rs:124]: This only works because the type is `Copy`, while the old example *deliberately* uses a non-Copy type. Please show how to init non-Copy arrays.

File: library/core/src/mem/maybe_uninit.rs [hunk 3]
```diff
@@ -1284,3 +1278,23 @@ impl<T> MaybeUninit<T> {
         }
     }
 }
+
+#[stable(feature = "foo", since = "CURRENT_RUSTC_VERSION")]
```
Comment (RalfJung) [inline library/core/src/mem/maybe_uninit.rs:1282]: Probably using `foo` here is what broke the other test.  Please use a proper feature name.

**Review by RalfJung — COMMENTED** [review]
The transmutation itself is sound. This is insta-stable though so definitely needs careful T-libs review.

File: library/core/src/mem/maybe_uninit.rs (hunk 2 above)
Comment (SUPERCILEX) [reply library/core/src/mem/maybe_uninit.rs:124]: Doesn't the `String` example right below show that?

File: library/core/src/mem/maybe_uninit.rs (hunk 3 above)
Comment (SUPERCILEX) [reply library/core/src/mem/maybe_uninit.rs:1282]: Doh, totally spaced on that, thanks.

File: library/core/src/mem/maybe_uninit.rs (hunk 2 above)
Comment (RalfJung) [reply library/core/src/mem/maybe_uninit.rs:124]: That doesn't initialize an entire array, it just works with a partially filled array. IMO we should have an example for actually crating a `[T; N]` for a non-Copy `T` via in-place init.

Comment (SUPERCILEX) [reply library/core/src/mem/maybe_uninit.rs:124]: Gotya, done.

File: library/core/src/mem/maybe_uninit.rs [hunk 4]
```diff
@@ -117,28 +117,42 @@ use crate::slice;
...
-///     let mut data: [MaybeUninit<Vec<u32>>; 1000] = unsafe {
-///         MaybeUninit::uninit().assume_init()
-///     };
+///     // Create an uninitialized array of `MaybeUninit`.
+///     let mut data = [MaybeUninit::uninit(); 1000];
 ///
-///     // Dropping a `MaybeUninit` does nothing, so if there is a panic during this loop,
-///     // we have a memory leak, but there is no memory safety issue.
 ///     for elem in &mut data[..] {
-///         elem.write(vec![42]);
```
Comment (RalfJung) [inline library/core/src/mem/maybe_uninit.rs:133]: Why did you delete the `Vec` example? Is there anything wrong with it?

It's certainly more plausible that a non-copy `u8` wrapper.

File: library/core/src/mem/maybe_uninit.rs (hunk 2 above)
Comment (RalfJung) [reply library/core/src/mem/maybe_uninit.rs:124]: The `Copy` example is also outright silly, since one could just write `[42; 1000]`. So in its current from I think we'd be better off without that example. If you know an actually useful example with a Copy type, feel free to use that.

Comment (RalfJung) [reply library/core/src/mem/maybe_uninit.rs:124]: Basically I think we'd be best off with exactly the old example, except that it uses the new transposition conversion. I'm open to improving the example but would like to know which issue is fixed by that improvement. :)

File: library/core/src/mem/maybe_uninit.rs (hunk 4 above)
Comment (SUPERCILEX) [reply library/core/src/mem/maybe_uninit.rs]: Yeah, fair enough, restored. I always thought the Vec example was weird but you're right this is no better.

File: library/core/src/mem/maybe_uninit.rs (hunk 2 above)
Comment (SUPERCILEX) [reply library/core/src/mem/maybe_uninit.rs]: Yup yup, done. :)

File: library/core/src/mem/maybe_uninit.rs [hunk 5]
```diff
@@ -135,10 +131,10 @@ use crate::slice;
 ///
 ///     // Everything is initialized. Transmute the array to the
 ///     // initialized type.
```
Comment (RalfJung) [inline library/core/src/mem/maybe_uninit.rs:133]:
```suggestion
///     // Everything is initialized. Convert the array to the initialized type.
```

File: library/core/src/mem/maybe_uninit.rs [hunk 6]
```diff
@@ -135,10 +131,10 @@ use crate::slice;
...
-///     unsafe { mem::transmute::<_, [Vec<u32>; 1000]>(data) }
+///     unsafe { MaybeUninit::<[Vec<_>; 1000]>::assume_init(data.into()) }
 /// };
 ///
-/// assert_eq!(&data[0], &[42]);
+/// assert_eq!(&data[888], &[42]);
```
Comment (RalfJung) [inline library/core/src/mem/maybe_uninit.rs:137]:
```suggestion
/// assert_eq!(&data[42], &[42]);
```
seems like the obvious choice ;)

**Comment by RalfJung** [comment]
Looks good, apart from some nits.

But I can't judge if `Into` instances are the right way to expose such APIs; I leave that to T-libs.

**Comment by SUPERCILEX** [comment]
Sweet, thanks for the review! :)

**Comment by scottmcm** [comment]
`impl`s are insta-stable, so this needs to go to libs-api.

@rustbot label +needs-fcp
r? rust-lang/libs-api

**Comment by scottmcm** [comment]
r? rust-lang/libs

**Comment by scottmcm** [comment]
...oh, I guess there's no way to get an -api member from the bot.  Well then how about

r? @m-ou-se

**Comment by SUPERCILEX** [comment]
@thomcc WDYT?

File: library/core/src/mem/maybe_uninit.rs [hunk 7]
```diff
@@ -1284,3 +1279,44 @@ impl<T> MaybeUninit<T> {
...
+    /// # use std::mem::MaybeUninit;
+    ///
+    /// let data: [MaybeUninit<u8>; 1000] = MaybeUninit::uninit().transpose();
+    /// ```
+    #[unstable(feature = "maybe_uninit_uninit_array_transpose", issue = "96097")]
+    pub fn transpose(self) -> [MaybeUninit<T>; N] {
+        // SAFETY: T and MaybeUninit<T> have the same layout
+        let result = unsafe { super::transmute_copy(&self) };
+        super::forget(self);
+        result
```
Comment (scottmcm) [inline library/core/src/mem/maybe_uninit.rs:1299]: suggestion: while this works fine, since none of it can panic, we try in the docs to follow the "pre-pit your peaches" order of things, rather than "forget it afterwards", and consider `forget` soft-deprecated.

So please update this to something like 
```rust
    pub fn transpose(self) -> [MaybeUninit<T>; N] {
        // SAFETY: T, ManuallyDrop<T>, and MaybeUninit<T> have the same layout
        unsafe { super::transmute_copy(&ManuallyDrop::new(self)) }
    }
```

(And ditto for the one below)

File: library/core/src/mem/maybe_uninit.rs [hunk 8]
```diff
@@ -117,41 +117,36 @@ use crate::slice;
...
 ///
 /// let data = {
-///     // Create an uninitialized array of `MaybeUninit`. The `assume_init` is
-///     // safe because the type we are claiming to have initialized here is a
-///     // bunch of `MaybeUninit`s, which do not require initialization.
-///     let mut data: [MaybeUninit<Vec<u32>>; 1000] = unsafe {
-///         MaybeUninit::uninit().assume_init()
-///     };
+///     // Create an uninitialized array of `MaybeUninit`.
+///     let mut data: [MaybeUninit<Vec<u32>>; 1000] = MaybeUninit::uninit().transpose();
```
Comment (scottmcm) [inline library/core/src/mem/maybe_uninit.rs:125]: fix: This is a great update, but we can't take it yet.  As `MaybeUninit` is a stable type, it's important that the type-level documentation says how to do things *on stable*.  Once `transpose` is stable then we should absolutely recommend it here, but for now it needs to give people something they can use without nightly.

So please leave this part of the docs as-is, but include an example of using `transpose` like this on the unstable method for nightly users to see.

(This is why `sort`'s documentation didn't mention `sort_unstable` until the latter was stabilized, etc.)

File: library/core/src/mem/maybe_uninit.rs (hunk 7 above)
Comment (SUPERCILEX) [reply library/core/src/mem/maybe_uninit.rs:1299]: Oh yeah, that's so much better. Is there a reason `forget` isn't deprecated? It seems like you could always replace it with `let _ = ManuallyDrop::new(foo);` (or `drop(ManuallyDrop::new(foo));`) worse case scenario.

File: library/core/src/mem/maybe_uninit.rs (hunk 8 above)
Comment (SUPERCILEX) [reply library/core/src/mem/maybe_uninit.rs]: Oops, good point. Done.

File: library/core/src/mem/maybe_uninit.rs (hunk 7 above)
Comment (scottmcm) [reply library/core/src/mem/maybe_uninit.rs]: Yeah, you can -- in fact that's literally what `forget` does

https://github.com/rust-lang/rust/blob/b8c35ca26b191bb9a9ac669a4b3f4d3d52d97fb1/library/core/src/mem/mod.rs#L153-L155

I think it's not deprecated because, **un**like `mem::uninitialized`, there's lots of code that used `forget` that's still perfectly correct because it only does things (like `copy_nonoverlapping`) that can't panic.

But maybe it's time to start thinking about it, now that `ManuallyDrop` has been around for long enough.  Or at least a deprecated-in-rust-2.0 for now so it gets picked up by https://github.com/rust-lang/rust/blob/9b0a099dfc9a97ecd10adb319396c731c4b2d169/compiler/rustc_lint_defs/src/builtin.rs#L2243-L2258

**Comment by scottmcm** [comment]
Thanks!  These look good for nightly to me.

@bors r+

File: library/core/src/mem/maybe_uninit.rs (hunk 7 above)
Comment (scottmcm) [reply library/core/src/mem/maybe_uninit.rs]: Hmm, maybe not.  It looks like `mem::forget(guard);` is still very common in the library for panic guards, where there's no need to consume them like in the example, and thus the `forget` phrasing is kinda nice.

But maybe that also just means that the `PanicGuard` types should have a `defuse` method or something as part of its API instead of telling people to `forget` it.

Comment (SUPERCILEX) [reply library/core/src/mem/maybe_uninit.rs]: Hmmm, yeah I see the appeal for panic guards, but poking around the stdlib I think it's actually pretty hard to understand what's going on unless you know what the guards are for, so maybe encouraging people to write a `defuse` method or add a comment explaining the weirdness is a good thing. Created #103101.

**Comment by RalfJung** [comment]
The PR title is outdated now, right?

**Comment by SUPERCILEX** [comment]
Yup, fixed.

**Comment by SimonSapin** [comment]
Is `transmute_copy(&ManuallyDrop::new(self))` different from `transmute(self)`?

**Comment by scottmcm** [comment]
@SimonSapin Yes, because `transpose` doesn't know that `MaybeUninit<[T; N]>` and `[MaybeUninit<T>; N]` are the same size, and thus fails to compile.

**Comment by thomcc** [comment]
> Yes, because `transpose`...

(I think you mean transmute, and only mention it because it might be pretty confusing with the mistake)
